core/data_flyout.js

```diff
@@ -0,0 +1,379 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2017 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Utility functions for handling variables.
+ * @author marisaleung@google.com (Marisa Leung)
+ */
+'use strict';
+
+/**
+ * @name Blockly.DataFlyout
+ * @namespace
+ **/
+goog.provide('Blockly.DataFlyout');
+
+goog.require('Blockly.Blocks');
+goog.require('Blockly.VariableModel');
+goog.require('Blockly.Variables');
+goog.require('Blockly.Workspace');
+
+/**
+ * Construct the blocks required by the flyout for the variable category.
+ * @param {!Blockly.Workspace} workspace The workspace contianing variables.
+ * @return {!Array.<!Element>} Array of XML block elements.
+ */
+Blockly.DataFlyout = function(workspace) {
+  var variableModelList = workspace.getVariablesOfType('');
```

Comment:
350 line function?  Decompose into multiple function (~1 per top-level block), and figure out if any of them share the same XML other than the name of the block.